File: crypto_algs/cpp/rsa.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cassert>

// Greatest Common Divisor
int gcd(int a, int b) {
    while (b != 0) {
        int tmp = b;
        b = a % b;
        a = tmp;
    }
    return a;
}

// Modular Exponentiation (base^exp % mod)
// Assumes larger numbers and applies mod iteratively.
long long mod_exp(long long base, long long exp, long long mod) {
    long long result = 1;
    // base^exp % mod = (base%mod)^exp % mod
    base %= mod;

    while (exp > 0) {
        // If odd, multiply by base once
        if (exp % 2 == 1)
            result = (result * base) % mod;
        // Remove odd factor and divide by two
        exp = exp >> 1;
        // Include the square in the answer
        base = (base * base) % mod;
    }
    return result;
}

// Modular Inverse using Extended Euclidean Algorithm
// Find co-prime numbers such that it returns d
// where e * d % phi == 1
int mod_inverse(int e, int phi) {
    int t = 0, newt = 1;
    int r = phi, newr = e;

    while (newr != 0) {
        int quotient = r / newr;
        int temp = newt;
        newt = t - quotient * newt;
        t = temp;

        temp = newr;
        newr = r - quotient * newr;
        r = temp;
    }

    if (r > 1) return -1; // e is not invertible
    if (t < 0) t += phi;
    return t;
}

// Encrypt string message to vector of ciphertext integers
std::vector<long long> encrypt_string(const std::string& message, int e, int n) {
    std::vector<long long> encrypted;
    for (char ch : message) {
        long long m = static_cast<int>(ch);
        encrypted.push_back(mod_exp(m, e, n));
    }
    return encrypted;
}

// Decrypt vector of ciphertext to string
std::string decrypt_string(const std::vector<long long>& encrypted, int d, int n) {
    std::string decrypted;
    for (long long c : encrypted) {
        long long m = mod_exp(c, d, n);
        decrypted += static_cast<char>(m);
    }
    return decrypted;
}

int main() {
    // Small primes
    int p = 61;
    int q = 53;

    // Compute n and phi(n)
    int n = p * q;
    int phi = (p - 1) * (q - 1);

    // Choose e (public key exponent)
    int e = 17;
    while (gcd(e, phi) != 1)
        ++e;

    // Compute d (private key exponent)
    int d = mod_inverse(e, phi);
    if (d == -1) {
        std::cerr << "Failed to find modular inverse.\n";
        return 1;
    }
    
    int answer = (e*d)%phi;
    printf("e*d %c n = %d\n", 37, answer);
    assert( (e*d) % phi == 1);

    std::cout << "Public key: (" << e << ", " << n << ")\n";
    std::cout << "Private key: (" << d << ", " << n << ")\n";

    std::string message = "HELLO RSA!";
    std::cout << "Original Message: " << message << "\n";

    // Encrypt: c = m^e mod n
    auto encrypted = encrypt_string(message, e, n);
    std::cout << "Encrypted: ";
    for (auto c : encrypted)
        std::cout << c << " ";
    std::cout << "\n";

    // Decrypt: m = c^d mod n
    std::string decrypted = decrypt_string(encrypted, d, n);
    std::cout << "Decrypted Message: " << decrypted << "\n";

    return 0;
}

File: crypto_algs/rsa.py
import random
import sympy
try:
    from .utils import sieve_of_eratosthenes
except ImportError:
    from utils import sieve_of_eratosthenes

LIMIT = 100
PRIMES_TO_LIMIT = sieve_of_eratosthenes(LIMIT)

def gcd(a, b):
    while b != 0:
        a, b = b, a % b
    return a

# def mod_inverse(e, phi):
#     d = 0
#     x1, x2, y1 = 0, 1, 1
#     temp_phi = phi
#     while e > 0:
#         temp1, temp2 = temp_phi // e, temp_phi - (temp_phi // e) * e
#         temp_phi, e = e, temp2
#         x, y = x2 - temp1 * x1, d - temp1 * y1
#         x2, x1, d, y1 = x1, x, y1, y
#         if temp_phi == 1:
#             return d + phi
def mod_inverse(e, phi):
    """
    Calculates the modular multiplicative inverse of e modulo phi.
    Returns d such that (d * e) % phi == 1.
    """
    g, x, y = extended_gcd(e, phi)
    if g != 1:
        raise Exception('Modular inverse does not exist')
    return x % phi

def extended_gcd(a, b):
    """
    Extended Euclidean Algorithm.
    Returns a tuple (g, x, y) such that a*x + b*y = g = gcd(a, b).
    """
    if a == 0:
        return b, 0, 1
    else:
        g, x, y = extended_gcd(b % a, a)
        return g, y - (b // a) * x, x
    
def encrypt(public_key, plaintext):
    key, n = public_key
    cipher = [(ord(char) ** key) % n for char in plaintext]
    return cipher

def decrypt(private_key, ciphertext):
    key, n = private_key
    plain = [chr((char ** key) % n) for char in ciphertext]
    return ''.join(plain)

def is_prime(num):
    if num <= 1:
        return False
    if num in PRIMES_TO_LIMIT:
        return True
    for p in PRIMES_TO_LIMIT:
        if num % p == 0:
            return False
    for i in range(LIMIT, int(num**0.5) + 1):
        if num % i == 0:
            return False
    return True

def find_next_prime(n):
    next_num = n + 2 if n%2 == 1 else n + 1
    while True:
        if is_prime(next_num):
            return next_num
        next_num += 1

def generate_keys():
    i_p = random.randint(2, len(PRIMES_TO_LIMIT)-3)
    i_q = random.randint(i_p+1, len(PRIMES_TO_LIMIT)-2)
    i_e = random.randint(i_q+1, len(PRIMES_TO_LIMIT)-1)
    p = PRIMES_TO_LIMIT[i_p]#sympy.nextprime(random.randint(5, 100))
    q = PRIMES_TO_LIMIT[i_q]#sympy.nextprime(random.randint(p+1, 200))
    e = PRIMES_TO_LIMIT[i_e]#sympy.nextprime(q)
    n = p * q
    phi = (p - 1) * (q - 1)
    print(e, phi)
    d = mod_inverse(e, phi)
    print(d)
    public_key = (e, n)
    private_key = (d, n)
    return public_key, private_key

if __name__ == "__main__":
    p = 53
    q = 59
    n = p * q
    phi = (p - 1) * (q - 1)
    e = 3
    d = mod_inverse(e, phi)

    public_key = (e, n)
    private_key = (d, n)

    message = "HELLO"
    encrypted_msg = encrypt(public_key, message)
    print("Encrypted message:", encrypted_msg)

    decrypted_msg = decrypt(private_key, encrypted_msg)
    print("Decrypted message:", decrypted_msg)